src/event.cc

```diff
@@ -1,325 +1,251 @@
-    }
+  Napi::Value mb = event.Get("mb");
+  if (!mb.IsObject() || mb.IsArray()) {
+    return false;
+  }
 
-    // supply the metadata for the event. oboe_event_init() will create a new
-    // random op ID for the event. (The op ID can be specified using the 3rd
-    // argument but there is no benefit to doing so.)
-    init_status = oboe_event_init(&this->event, &omd, NULL);
-    if (init_status != 0) {
-      Napi::Error::New(env, "oboe.event_init: " + std::to_string(init_status)).ThrowAsJavaScriptException();
-      return;
-    }
+  Napi::Value buf = mb.As<Napi::Object>().Get("buf");
+  if (!buf.IsBuffer()) {
+    return false;
+  }
+  Napi::Buffer<uint8_t> b = buf.As<Napi::Buffer<uint8_t>>();
+  if (b.Length() != 30) {
+    return false;
+  }
 
-    if (add_edge) {
-      int edge_status = oboe_event_add_edge(&this->event, &omd);
-      if (edge_status != 0) {
-        Napi::Error::New(env, "oboe.add_edge: " + std::to_string(edge_status)).ThrowAsJavaScriptException();
-        return;
-      }
-    }
+  return true;
 }
 
 //
-// C++ callable function to create a JavaScript Event object.
+// initialize oboe_metadata_t structure from a Metabuf buffer
 //
-Napi::Object Event::NewInstance(Napi::Env env) {
-    Napi::EscapableHandleScope scope(env);
+void initializeOboeMd(oboe_metadata_t &md, Napi::Buffer<uint8_t> b) {
+  md.version = XTR_CURRENT_VERSION;
+
+  // copy IDs
+  const uint kTaskIdBase = 1;
+  const uint kOpIdBase = kTaskIdBase + OBOE_MAX_TASK_ID_LEN;
+  for (uint i = 0; i < OBOE_MAX_TASK_ID_LEN; i++) {
+    md.ids.task_id[i] = b[kTaskIdBase + i];
+  }
+  for (uint i = 0; i < OBOE_MAX_OP_ID_LEN; i++) {
+    md.ids.op_id[i] = b[kOpIdBase + i];
+  }
 
-    Napi::Object o = constructor.New({});
+  // initialize lengths
+  md.task_len = OBOE_MAX_TASK_ID_LEN;
+  md.op_len = OBOE_MAX_OP_ID_LEN;
 
-    return scope.Escape(napi_value(o)).ToObject();
+  // and finally flags.
+  md.flags = b[kOpIdBase + OBOE_MAX_OP_ID_LEN];
 }
 
 //
-// C++ callable function to create a JavaScript Event object.
+// add KV pair
 //
-// This signature includes an optional boolean for whether an edge
-// should be set or not. It defaults to true.
+// returns -1 if error.
 //
-Napi::Object Event::NewInstance(Napi::Env env, oboe_metadata_t* omd, bool edge) {
-  Napi::EscapableHandleScope scope(env);
+int addKvPair(oboe_event_t* event, std::string key, Napi::Value value) {
+  int status = -1;
 
-  Napi::Value v0 = Napi::External<oboe_metadata_t>::New(env, omd);
-  Napi::Value v1 = Napi::Boolean::New(env, edge);
+  if (value.IsBoolean()) {
+    status = oboe_event_add_info_bool(event, key.c_str(), value.As<Napi::Boolean>().Value());
+  } else if (value.IsNumber()) {
+    const double v = value.As<Napi::Number>();
 
-  Napi::Object o = constructor.New({v0, v1});
-
-  return scope.Escape(napi_value(o)).ToObject();
-}
+    // it it's within integer range add an integer
+    if (v > -(pow(2, 53) - 1) && v < pow(2, 53) -1) {
```

Comment:
good questions:
1) INT_MAX - i can't use because it's the largest integer than can be represented by a floating point number. one of javascript's bit limitations is that it doesn't have an integer type - everything is floating point. so pow(2, 53) - 1 isn't in limits.h
2) i guess i could; i made the presumption that gcc would optimize away the calculation (and it does) but it's probably more elegant to give it a name as opposed to making someone read it twice. i'll change it.
3) v is not an integer; it's a double. i think c++ will convert a double to an integer without complaint which is why the check for the limit is required - it's not really an integer to node unless it is smaller than the pow(2, 53) - 1.
4) no, i didn't mean "it it" - thanks.

Reply:
i'm changing the logic to be `>=` and `<=`. i think i was off by one on both.